File: LeetCode/Tree/Validate_Binary_Search_Tree.cpp
/*Given a binary tree, determine if it is a valid binary search tree (BST).

Assume a BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
*/

#include <iostream>
#include <stdlib.h>
#include <limits.h>
using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x):val(x),left(NULL),right(NULL){}
};

class Solution{
public:
    bool isValidBST(TreeNode *root){
        return isValid(root,INT_MIN,INT_MAX);
    }
    bool isValid(TreeNode *node,int MIN,int MAX){
        if(node==NULL){
            return true;
        }
        if(node->val>MIN&&node->val<MAX
                &&isValid(node->left,MIN,node->val)
                &&isValid(node->right,node->val,MAX)){
            return true;
        }else{
            return false;
        }
    }
};

TreeNode *newNode(int data){
    TreeNode *node=(TreeNode*)malloc(sizeof(TreeNode));
    node->val=data;
    node->left=NULL;
    node->right=NULL;
}

int main(){
    Solution solution;
    TreeNode *root=newNode(3);
    root->left=newNode(2);
    root->right=newNode(4);
    cout<<solution.isValidBST(root)<<endl; //1
    return 0;
}

File: LeetCode_Python/Tree/Validate_Binary_Search_Tree.py
'''
Given a binary tree, determine if it is a valid binary search tree (BST).

Assume a BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.

Example 1:

    2
   / \
  1   3

Binary tree [2,1,3], return true.

Example 2:

    1
   / \
  2   3

Binary tree [1,2,3], return false.
'''

'''
class TreeNode(object):
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None
'''

class Solution(object):
	def isValidBST(self, root):
		INT_MAX = 4294967296
		INT_MIN = -4294967296
		def dfs(root, min, max):
			if root == None:
				return True 
			if root.val <= min or root.val >= max:
				return False 
			return dfs(root.left, min, root.val) and dfs(root.right, root.val, max)
		return dfs(root, INT_MIN, INT_MAX)
